cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/AppFabricServer.java

```diff
@@ -110,36 +102,31 @@ public class AppFabricServer extends AbstractIdleService {
-                         RunRecordCorrectorService runRecordCorrectorService,
-                         ProgramRunStatusMonitorService programRunStatusMonitorService,
-                         ApplicationLifecycleService applicationLifecycleService,
-                         ProgramNotificationSubscriberService programNotificationSubscriberService,
-                         ProgramStopSubscriberService programStopSubscriberService,
-                         @Named("appfabric.services.names") Set<String> servicesNames,
-                         @Named("appfabric.handler.hooks") Set<String> handlerHookNames,
-                         AuditLogSubscriberService auditLogSubscriberService,
-                         CoreSchedulerService coreSchedulerService,
-                         CredentialProviderService credentialProviderService,
-                         NamespaceCredentialProviderService namespaceCredentialProviderService,
-                         ProvisioningService provisioningService,
-                         BootstrapService bootstrapService,
-                         SystemAppManagementService systemAppManagementService,
-                         TransactionRunner transactionRunner,
-                         RunRecordMonitorService runRecordCounterService,
-                         CommonNettyHttpServiceFactory commonNettyHttpServiceFactory,
-                         RunDataTimeToLiveService runDataTimeToLiveService,
-                         SourceControlOperationRunner sourceControlOperationRunner,
-                         RepositoryCleanupService repositoryCleanupService,
-                         OperationNotificationSubscriberService operationNotificationSubscriberService) {
+      DiscoveryService discoveryService,
+      @Named(Constants.Service.MASTER_SERVICES_BIND_ADDRESS) InetAddress hostname,
+      @Named(Constants.AppFabric.SERVER_HANDLERS_BINDING) Set<HttpHandler> handlers,
+      @Nullable MetricsCollectionService metricsCollectionService,
+      ProgramRuntimeService programRuntimeService,
+      ProgramRunStatusMonitorService programRunStatusMonitorService,
+      ApplicationLifecycleService applicationLifecycleService,
+      @Named("appfabric.services.names") Set<String> servicesNames,
+      @Named("appfabric.handler.hooks") Set<String> handlerHookNames,
+      CoreSchedulerService coreSchedulerService,
+      CredentialProviderService credentialProviderService,
+      NamespaceCredentialProviderService namespaceCredentialProviderService,
+      ProvisioningService provisioningService,
+      BootstrapService bootstrapService,
+      SystemAppManagementService systemAppManagementService,
+      TransactionRunner transactionRunner,
+      RunRecordMonitorService runRecordCounterService,
+      CommonNettyHttpServiceFactory commonNettyHttpServiceFactory,
+      SourceControlOperationRunner sourceControlOperationRunner,
+      RepositoryCleanupService repositoryCleanupService) {
```

Comment:
What's about SCM. It has persistent state. Do we have it on all AppFabricServer instances? It will be cleaned on each restart, won't it?

Reply:
`SourceControlOperationRunner` is required for `SourceControlManagementService`, which is used by `SourceControlManagementHttpHandler`.

`RepositoryCleanupService` needs to run in the same pod as `SourceControlOperationRunner`.

Keeping them in the Server, and not in the processor.

I understand that it caches the repositories in /tmp/source-control/. But it is ok if the cache gets cleaned up on restart.

Reply:
OK, I just want to be sure we understand the performance hit. BTW: Are we using in memory or worker based runner? In the latter case I think it's already on workers

Reply:
Confirmed with @samdgupi and @adrikagupta :

1. For multi push/pull Appfabric (`InMemorySourceControlOperationRunner`) is used and for other operations Task workers (`RemoteSourceControlOperationRunner`) are used if task workers are enabled.

2. It is safe to delete the repositories on Pod restarts. Repositories are cleaned after each operation: `io.cdap.cdap.sourcecontrol.RepositoryManager#close`. `RepositoryCleanupService` ensures that leaked directories will get removed periodically.

3. A unique directory is created per operation. So cleaning up repositories on restart won't have performance impact. 

4. Not for SCM directly but for LRO's we store `OperationControllers` in `InMemoryOperationRuntime` but they don't need to get persisted across restarts.